web/src/main/webapp/lamp.ts

```diff
@@ -0,0 +1,412 @@
+import {Dialog} from "./dialog";
+import {TopMenu, TopSubMenu} from "./menu";
+import {TableDataView, TableView, TableRenderer} from "./table";
+import {ContentsKind, RangeInfo, BasicColStats, Schema, RemoteTableObject, RemoteTableObjectView, RemoteTableRenderer, ColumnAndRange, RecordOrder} from "./tableData";
+import {FullPage, IDataView, Resolution} from "./ui";
+import {Renderer, RpcReceiver, RpcRequest} from "./rpc";
+import {PartialResult, Point2D, clamp, Pair} from "./util";
+import {HeatMapData} from "./heatMap";
+import {HeatMapArrayDialog} from "./heatMapArray";
+import {ColorMap, ColorLegend} from "./vis";
+import d3 = require('d3');
+
+export class PointSet2D {
+    points: Point2D[];
+}
+
+class ControlPointsView extends RemoteTableObjectView {
+    private minX: number;
+    private minY: number;
+    private maxX: number;
+    private maxY: number;
+    private heatMapCanvas: any;
+    private heatMapChart: any;
+    private controlPointsCanvas: any;
+    private controlPointsChart: any;
+    public controlPoints: PointSet2D;
+    private heatMap: HeatMapData;
+    private lampTableObject: RemoteTableObject;
+    private colorMap: ColorMap;
+    private colorLegend: ColorLegend;
+
+    constructor(private originalTableObject: RemoteTableObject, page: FullPage, private controlPointsId, private selectedColumns) {
+        super(originalTableObject.remoteObjectId, page);
+        this.topLevel = document.createElement("div");
+        this.topLevel.classList.add("chart");
+        this.topLevel.classList.add("lampView");
+
+        let menu = new TopMenu( [
+            { text: "View", subMenu: new TopSubMenu([
+                { text: "refresh", action: () => { this.refresh(); } },
+                { text: "update ranges", action: () => { this.fetchNewRanges(); } },
+                { text: "table", action: () => { this.showTable(); } },
+                { text: "3D heat map", action: () => { this.heatMap3D(); } },
+            ]) }
+        ]);
+
+        this.topLevel.appendChild(menu.getHTMLRepresentation());
+
+        this.colorMap = new ColorMap(0, 1);
+        this.colorLegend = new ColorLegend(this.colorMap);
+        this.colorLegend.setColorMapChangeEventListener(() => {
+            this.refresh();
+        });
+        this.topLevel.appendChild(this.colorLegend.getHTMLRepresentation());
+        let chartDiv = document.createElement("div")
+        this.topLevel.appendChild(chartDiv);
+
+        let canvasSize = Math.min(Resolution.getCanvasSize(this.getPage()).width, Resolution.getCanvasSize(this.getPage()).height);
+        let chartSize = Math.min(Resolution.getChartSize(this.getPage()).width, Resolution.getChartSize(this.getPage()).height)
```

Comment:
You should use semicolons at the end of the statements; makes the code easier to read.

Reply:
I really dislike optional semicolons in languages. Either require them, or don't allow them when they're redundant! ;-)